### PR TITLE
Make liquidations work with any asset

### PR DESCRIPTION
**Pull request by MiksuJak**

### CODE REVIEW

File: contracts/truefi2/Liquidator2.sol [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    }
+
+    /**
+     * @dev Liquidates a defaulted Loan, withdraws a portion of tru from staking pool
+     * then transfers tru to TrueFiPool as compensation
+     * @param loan Loan to be liquidated
+     */
+    function liquidate(ILoanToken2 loan) external {
+        require(loanFactory.isLoanToken(address(loan)), "Liquidator: Unknown loan");
+        ITrueFiPool2 pool = ITrueFiPool2(loan.pool());
```
Comment (hal909) [inline contracts/truefi2/Liquidator2.sol:89]: check here that loan.token() is on approvedTokens

Comment (hal909) [reply contracts/truefi2/Liquidator2.sol:89]: require(approvedLoans(pool.token() != 0), "Liquidator: Token not approved for default protection");

File: contracts/truefi2/Liquidator2.sol [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    // ===== THIS CONTRACT IS INITIALIZABLE ======
+    // === STORAGE VARIABLES ARE DECLARED BELOW ==
+    // REMOVAL OR REORDER OF VARIABLES WILL RESULT
+    // ========= IN STORAGE CORRUPTION ===========
+
+    IPoolFactory public poolFactory;
+    IStakingPool public stkTru;
+    IERC20 public tru;
+    ILoanFactory2 public loanFactory;
+
```
Comment (hal909) [inline contracts/truefi2/Liquidator2.sol:34]: Add approved tokens mapping

mapping(address => bool) approvedTokens;

File: contracts/truefi2/Liquidator2.sol [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+     * @dev Set new max fetch share
+     * @param newShare New share to be set
+     */
+    function setFetchMaxShare(uint256 newShare) external onlyOwner {
+        require(newShare > 0, "Liquidator: Share cannot be set to 0");
+        require(newShare <= 10000, "Liquidator: Share cannot be larger than 10000");
+        fetchMaxShare = newShare;
+        emit FetchMaxShareChanged(newShare);
+    }
+
```
Comment (hal909) [inline contracts/truefi2/Liquidator2.sol:81]: Add function setTokenApproval(address _token) onlyOwner which allows setting token approvals in approvedTokens

File: contracts/truefi2/Liquidator2.sol [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+     * @dev Liquidates a defaulted Loan, withdraws a portion of tru from staking pool
+     * then transfers tru to TrueFiPool as compensation
+     * @param loan Loan to be liquidated
+     */
+    function liquidate(ILoanToken2 loan) external {
+        require(loanFactory.isLoanToken(address(loan)), "Liquidator: Unknown loan");
+        ITrueFiPool2 pool = ITrueFiPool2(loan.pool());
+        uint256 defaultedValue = getAmountToWithdraw(loan.debt().sub(loan.repaid()), ITrueFiPoolOracle(pool.oracle()));
+        stkTru.withdraw(defaultedValue);
+        require(loan.status() == ILoanToken2.Status.Defaulted, "Liquidator: Loan must be defaulted");
```
Comment (hal909) [inline contracts/truefi2/Liquidator2.sol:92]: check-effects: move this require to the beginning of the function before making the withdrawal

File: contracts/truefi2/Liquidator2.sol [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+        loan.liquidate();
+        require(tru.transfer(address(pool), defaultedValue));
+        emit Liquidated(loan);
```
Comment (hal909) [inline contracts/truefi2/Liquidator2.sol:95]: emit defaultedValue in this event

File: contracts/truefi2/Liquidator2.sol [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+     * @dev Emitted fetch max share is changed
+     * @param newShare New share set
+     */
+    event FetchMaxShareChanged(uint256 newShare);
+
+    /**
+     * @dev Emitted when a loan gets liquidated
+     * @param loan Loan that has been liquidated
+     */
+    event Liquidated(ILoanToken2 loan);
```
Comment (hal909) [inline contracts/truefi2/Liquidator2.sol:51]: add uint256 defaultedValue in this event

File: contracts/truefi2/Liquidator2.sol [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+    }
+
+    /**
+     * @dev Calculate amount of tru to be withdrawn from staking pool (not more than preset share)
+     * @param deficit Amount of tusd lost on defaulted loan
+     * @return amount of TRU to be withdrawn on liquidation
+     */
+    function getAmountToWithdraw(uint256 deficit, ITrueFiPoolOracle oracle) internal view returns (uint256) {
+        uint256 stakingPoolSupply = stkTru.stakeSupply();
+        uint256 maxWithdrawValue = stakingPoolSupply.mul(fetchMaxShare).div(10000);
```
Comment (hal909) [inline contracts/truefi2/Liquidator2.sol:105]: divide by BASIS constant

File: contracts/truefi2/Liquidator2.sol [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+        fetchMaxShare = 1000;
+    }
+
+    /**
+     * @dev Set new max fetch share
+     * @param newShare New share to be set
+     */
+    function setFetchMaxShare(uint256 newShare) external onlyOwner {
+        require(newShare > 0, "Liquidator: Share cannot be set to 0");
+        require(newShare <= 10000, "Liquidator: Share cannot be larger than 10000");
```
Comment (hal909) [inline contracts/truefi2/Liquidator2.sol:77]: check newShare <= BASIS instead of 10000

File: contracts/truefi2/Liquidator2.sol [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+    IStakingPool public stkTru;
+    IERC20 public tru;
+    ILoanFactory2 public loanFactory;
+
+    // max share of tru to be taken from staking pool during liquidation
+    // 1000 -> 10%
+    uint256 public fetchMaxShare;
+
+    // ======= STORAGE DECLARATION END ============
+
```
Comment (hal909) [inline contracts/truefi2/Liquidator2.sol:40]: add uint256 constant BASIS = 10000

**Review by hal909 — CHANGES_REQUESTED** [review]
A few things to change, most importantly adding a whitelist of tokens approved for liquidation. The whitelist should hold the token address (e.g. DAI address) and allow owner to add tokens approved for liquidation.

File: contracts/truefi2/Liquidator2.sol [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+    // basis point for ratio
+    uint256 private constant BASIS_RATIO = 10000;
+
+    // ================ WARNING ==================
+    // ===== THIS CONTRACT IS INITIALIZABLE ======
+    // === STORAGE VARIABLES ARE DECLARED BELOW ==
+    // REMOVAL OR REORDER OF VARIABLES WILL RESULT
+    // ========= IN STORAGE CORRUPTION ===========
+
+    IPoolFactory public poolFactory;
```
Comment (vanruch) [inline contracts/truefi2/Liquidator2.sol:33]: I think poolFactory is unused

File: contracts/truefi2/Liquidator2.sol [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+    /**
+     * @dev Liquidates a defaulted Loan, withdraws a portion of tru from staking pool
+     * then transfers tru to TrueFiPool as compensation
+     * @param loan Loan to be liquidated
+     */
+    function liquidate(ILoanToken2 loan) external {
+        require(loanFactory.isLoanToken(address(loan)), "Liquidator: Unknown loan");
+        require(loan.status() == ILoanToken2.Status.Defaulted, "Liquidator: Loan must be defaulted");
+        ITrueFiPool2 pool = ITrueFiPool2(loan.pool());
+        require(approvedTokens[address(pool.token())] == true, "Liquidator: Token not approved for default protection");
```
Comment (vanruch) [inline contracts/truefi2/Liquidator2.sol:115]:
```suggestion
        require(approvedTokens[address(pool.token())], "Liquidator: Token not approved for default protection");
```

**Review by hal909 — APPROVED** [review]

**Pull request merged**